includes/abstracts/abstract-wc-product.php

```diff
@@ -1374,13 +1376,22 @@ public function save() {
 		 */
 		do_action( 'woocommerce_before_' . $this->object_type . '_object_save', $this, $this->data_store );
 
+		$state = $this->before_data_store_save_or_update();
+
 		if ( $this->get_id() ) {
+			$changeset = $this->get_changes();
 			$this->data_store->update( $this );
 		} else {
+			$changeset = null;
 			$this->data_store->create( $this );
 		}
 
-		$this->maybe_defer_product_sync();
+		$this->after_data_store_save_or_update( $state );
+
+		// Update attributes lookup table if the product is new OR it's not but there are actually any changes.
+		if ( is_null( $changeset ) || ! empty( $changeset ) ) {
```

Comment:
A little confused here. Aren't these two conditions contradicting? If it is null, it will never be not empty and vice versa?

Reply:
The `if` evaluates to false only if the changeset is an empty array and that's on purpose. I want the condition to be true when the product is new (so `null` changeset) or if there are changes (non-`null` and non-empty changeset), but I want it to be false when the product exists and there are no changes (so the changeset is an empty array).

Reply:
Do you think I could update the comment so that the intent is more clear?